src/compiler/checker.ts

```diff
@@ -20287,11 +20289,18 @@ namespace ts {
             return declaredType;
         }
 
-        function getTypeFactsOfTypes(types: Type[]): TypeFacts {
+        function getTypeFactsOfTypes(types: Type[], isUnion: boolean): TypeFacts {
             let result: TypeFacts = TypeFacts.None;
             for (const t of types) {
                 result |= getTypeFacts(t);
             }
+            if (!isUnion) {
+                // Get the set of positive facts for the intersection by masking with the negative set shifted left,
+                // then shift those present positive facts into the negative fact value range, and unset any of those
+                // bits (by negating that mask and then intersecting it with the original value)

```

Comment:
I don't like "intersecting" here, but "ANDing" sounds silly too. I think there's a technical term, but it might be too technical because I can't remember it.

Consider my vague dislike registered and continue as appropriate.